clickhouse/query.go

```diff
@@ -0,0 +1,357 @@
+
+	// ErrTooBigResult means that daemon would probably exceed maximum response size if full result is sent.
+	ErrTooBigResult = fmt.Errorf("Result exceeds %d MiB", maxResultSize/(1<<20))
+
+	// ErrTooBigResponse means that ClickHouse server was going to send too big of a response to INSERT
+	ErrTooBigResponse = fmt.Errorf("Too big response from server")
+
+	// ErrTemporarilyUnavailable means that there are no servers alive that have the requested table
+	ErrTemporarilyUnavailable = errors.New("No servers available")
+
+	tableDstMap struct {
+		sync.RWMutex
+		m map[string]*destination.Setting
+	}
+
+	tableDstSessionIdsMap struct {
+		sync.RWMutex
+		nextGc time.Time
+		m      map[string]dstSessionIdItem
+	}
+
+	queryRegex      = regexp.MustCompile(`(?s)(?:SELECT|select)\s.*?(?:FROM|from)\s+([a-zA-Z_][a-zA-Z0-9_.]*(?:@[0-9]+)?)`)
+	queryHTTPClient = &http.Client{
+		Transport: &http.Transport{
+			MaxIdleConnsPerHost: 1,
+			DialContext:         srvfunc.CachingDialer,
+		},
+	}
+)
+
+func maybeAddTable(m map[string]*destination.Setting, table string, dst *destination.Setting) {
+	if _, ok := m[table]; !ok {
+		m[table] = dst
+	}
+}
+
+// Escape escapes string for MySQL. It should work for ClickHouse as well.
+func Escape(txt string) string {
+	var (
+		esc string
+		buf bytes.Buffer
+	)
+	last := 0
+	for ii, bb := range txt {
+		switch bb {
+		case 0:
+			esc = `\0`
+		case '\n':
+			esc = `\n`
+		case '\r':
+			esc = `\r`
+		case '\\':
+			esc = `\\`
+		case '\'':
+			esc = `\'`
+		case '"':
+			esc = `\"`
+		case '\032':
+			esc = `\Z`
+		default:
+			continue
+		}
+		io.WriteString(&buf, txt[last:ii])
+		io.WriteString(&buf, esc)
+		last = ii + 1
+	}
+	io.WriteString(&buf, txt[last:])
+	return buf.String()
+}
+
+// UpdateDestinationsConfig is used for QueryDeadline to determine where to send query.
+func UpdateDestinationsConfig(m destination.Map) {
+	tableDstMap.Lock()
+	defer tableDstMap.Unlock()
+
+	tableDstMap.m = make(map[string]*destination.Setting)
+
+	for _, settings := range m {
+		if settings.Default {
+			tableDstMap.m[defaultTable] = settings
+			continue
+		}
+
+		for _, tbl := range settings.Tables {
+			tableDstMap.m[tbl] = settings
+			maybeAddTable(tableDstMap.m, strings.TrimSuffix(tbl, "_buffer"), settings)
+			if !strings.Contains(tbl, ".") {
+				maybeAddTable(tableDstMap.m, "default."+tbl, settings)
+				maybeAddTable(tableDstMap.m, "default."+strings.TrimSuffix(tbl, "_buffer"), settings)
+			}
+		}
+	}
+}
+
+// GetDestinationSetting returns setting for specified table name
+func GetDestinationSetting(tableName string) *destination.Setting {
+	tableDstMap.RLock()
+	dst, ok := tableDstMap.m[tableName]
+	if !ok {
+		dst = tableDstMap.m[defaultTable]
+	}
+	tableDstMap.RUnlock()
+	return dst
+}
+
+func GetDestinationHostBySessionId(tableName string, params *QueryParams) (srv destination.ServerHostPort, ok bool) {
+	const gcEvery = 5 * time.Minute
+
+	// GC
```

Comment:
Why not just create a separate GC goroutine and remove this racy logic altogether?